File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.java
import java.util.*;

class Solution {
    
    static class Bus implements Comparable<Bus>{
        int time;
        int lastArrivalTime;
        boolean isfull=false;
        
        Bus(int time){
            this.time=time;
        }
        
        @Override
        public int compareTo(Bus o){
            return this.time - o.time;
        }
        
    }
    
    public int calTime(int idx, int min){
        int total = 540+min*idx;
        int result = total/60*100+total%60;
        return result;
    }
    
    
    public int subTime(int time){
        int total = (time%100+time/100*60)-1;
        int result = total/60*100+total%60;
        return result;
    }
    
    public String timeToString(int time){
        int hour = time/100;
        int min = time%100;
        return String.format("%02d:%02d", hour, min);
    }
    
    public String solution(int n, int t, int m, String[] timetable) {
        Bus[] busTimes = new Bus[n];
        int[] arrivalTimes = new int[timetable.length];
        
        for(int i=0; i<timetable.length; i++){
            int arrivalTime = Integer.parseInt(timetable[i].replace(":",""));
            arrivalTimes[i] = arrivalTime;
        }
        
        Arrays.sort(arrivalTimes);
        
        for(int i=0; i<n; i++){
            int busTime = calTime(i, t);
            busTimes[i] = new Bus(busTime);
        }
        
        int j=0;
        for(int i=0; i<n; i++){
            int busTime = busTimes[i].time;
            int cnt=0;
            
            for( ;j<timetable.length; j++){
                int arrivalTime = arrivalTimes[j];
                if(arrivalTime>busTime) break;
                cnt++;
                busTimes[i].lastArrivalTime=arrivalTime;
                if(cnt==m){
                    busTimes[i].isfull = true;
                    j++;
                    break;
                }
            }
        }
        
        Bus lastBus = busTimes[n-1];
        if(lastBus.isfull){
          return timeToString(subTime(lastBus.lastArrivalTime));
        }
        
        return timeToString(lastBus.time);
    }
}

File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.js
function solution(n, t, m, timetable) {
    // 크루 대기 시간 오름차순 정렬
    let timetables = timetable.map((e)=>{
        let time = parseInt(e.replace(":", ""), 10);
        return Math.floor(time/100)*60+time%100;
    }).sort((a,b)=>a-b);
    
    // 셔틀 시간
    let busTimes = [];
    for(let i=0; i<n; i++){
        busTimes.push(540+t*i);
    }
    
    // 각 셔틀에 크루들 태우기
    let lastBoardTime = 0;
    let isfull = false;
    for(busTime of busTimes){
        let cnt = 0;
        isfull = false;
        
        while(timetables.length>0){
            let arrivalTime = timetables.shift();
            if(arrivalTime>busTime) {
                timetables.unshift(arrivalTime);
                break;
            }
            cnt++;
            if(cnt===m){
                lastBoardTime = arrivalTime;
                isfull = true;
                break;
            }
            
        }
    }
    
    function stringFormat(time){
        let hour = Math.floor((time)/60);
        let min = (time)%60;
        return `${String(hour).padStart(2, "0")}:${String(min).padStart(2,"0")}`;
    }
    
    // 결론
    // 마지막 셔틀 만차
        //마지막 크루보다 1분 일찍 도착
    if(isfull){
        return stringFormat(lastBoardTime-1);
    }
    // 만차x
        // 마지막 셔틀 시간에 도착
    return stringFormat(busTimes[n-1]);
    
    
}